src/Framework/EncodingUtilities.cs

```diff
@@ -247,5 +251,88 @@ internal static Encoding BatchFileEncoding(string contents, string encodingSpeci
                         : EncodingUtilities.Utf8WithoutBom;
             }
         }
+
+        /// <summary>
+        /// The .NET SDK and Visual Studio both have environment variables that set a custom language. MSBuild should respect the SDK variable.
+        /// To use the correspoding UI culture, in certain cases the console encoding must be changed. This function will change the encoding in these cases.
+        /// This code introduces a breaking change in .NET 8 due to the encoding of the console being changed.
+        /// If the environment variables are undefined, this function should be a no-op.
+        /// </summary>
+        /// <returns>
+        /// The custom language that was set by the user for an 'external' tool besides MSBuild.
+        /// Returns <see langword="null"/> if none are set.
+        /// </returns>
+        public static CultureInfo GetExternalOverridenUILanguageIfSupportableWithEncoding()
```

Comment:
Should this be behind a changewave so it works by default but can be easily disabled if it causes an unforseen problem?

Reply:
You mean like an environment variable? Good idea, here we are: `MSBUILDDISABLEDYNAMICUTFENCODING`
![image](https://github.com/dotnet/msbuild/assets/23152278/0910b008-e08a-4cfb-ae9f-0bdc4607f446)

Reply:
We used to use that approach but it littered a bunch of environment variables around for all time. Now, we have a more systematic approach, detailed in https://github.com/dotnet/msbuild/blob/1dde6006e118eac5fc799ec8ad3f93b2089a2c57/documentation/wiki/ChangeWaves-Dev.md#condition-your-feature-on-a-change-wave -- can you use that instead? Since you merged `main` recently you should have access to `Wave17_8`.

Reply:
Done!

Reply:
Yeah, I sent ben this thread after seeing he wrote it (but he's on vacation for  a while)

Reply:
👋 glad it was helpful!